p2p/discover/table.go

```diff
@@ -600,177 +507,223 @@ func (tab *Table) filterNode(n *node) bool {
-// bucket. If the node is already in the bucket, it is moved to the front. If the bucket
-// has no space, the node is added to the replacements list.
-//
-// There is an additional safety measure: if the table is still initializing the node
-// is not added. This prevents an attack where the table could be filled by just sending
-// ping repeatedly.
-//
-// The caller must not hold tab.mutex.
-
-func (tab *Table) addVerifiedNode(n *node) {
-	gopool.Submit(func() {
-		tab.addVerifiedNodeSync(n)
-	})
+func (tab *Table) addIP(b *bucket, ip netip.Addr) bool {
+	if !ip.IsValid() || ip.IsUnspecified() {
+		return false // Nodes without IP cannot be added.
+	}
+	if netutil.AddrIsLAN(ip) {
+		return true
+	}
+	if !tab.ips.AddAddr(ip) {
+		tab.log.Debug("IP exceeds table limit", "ip", ip)
+		return false
+	}
+	if !b.ips.AddAddr(ip) {
+		tab.log.Debug("IP exceeds bucket limit", "ip", ip)
+		tab.ips.RemoveAddr(ip)
+		return false
+	}
+	return true
 }
 
-func (tab *Table) addVerifiedNodeSync(n *node) {
-	if !tab.isInitDone() {
+func (tab *Table) removeIP(b *bucket, ip netip.Addr) {
+	if netutil.AddrIsLAN(ip) {
 		return
 	}
-	if n.ID() == tab.self().ID() {
-		return
+	tab.ips.RemoveAddr(ip)
+	b.ips.RemoveAddr(ip)
+}
+
+// handleAddNode adds the node in the request to the table, if there is space.
+// The caller must hold tab.mutex.
+func (tab *Table) handleAddNode(req addNodeOp) bool {
+	if req.node.ID() == tab.self().ID() {
+		return false
 	}
-	if tab.filterNode(n) {
-		return
+
+	// For nodes from inbound contact, there is an additional safety measure: if the table
+	// is still initializing the node is not added.
+	if req.isInbound && !tab.isInitDone() {
+		return false
 	}
-	tab.mutex.Lock()
-	defer tab.mutex.Unlock()
-	b := tab.bucket(n.ID())
-	if tab.bumpInBucket(b, n) {
-		// Already in bucket, moved to front.
-		return
+
+	b := tab.bucket(req.node.ID())
+	n, _ := tab.bumpInBucket(b, req.node, req.isInbound)
+	if n != nil {
+		// Already in bucket.
+		return false
 	}
-	if len(b.entries) >= tab.bucketSize {
+	if len(b.entries) >= bucketSize {
 		// Bucket full, maybe add as replacement.
-		tab.addReplacement(b, n)
-		return
+		tab.addReplacement(b, req.node)
+		return false
 	}
-	if !tab.addIP(b, n.IP()) {
+	if !tab.addIP(b, req.node.IPAddr()) {
 		// Can't add: IP limit reached.
-		return
+		return false
 	}
 
-	// Add to front of bucket.
-	b.entries, _ = pushNode(b.entries, n, tab.bucketSize)
-	b.replacements = deleteNode(b.replacements, n)
-	n.addedAt = time.Now()
-
-	if tab.nodeAddedHook != nil {
-		tab.nodeAddedHook(b, n)
+	// Add to bucket.
+	wn := &tableNode{Node: req.node}
+	if req.forceSetLive {
+		wn.livenessChecks = 1
+		wn.isValidatedLive = true
 	}
-}
 
-// delete removes an entry from the node table. It is used to evacuate dead nodes.
-func (tab *Table) delete(node *node) {
-	tab.mutex.Lock()
-	defer tab.mutex.Unlock()
+	// TODO(Matus): fix the filterNode feature
```

Comment:
p2p module changes a lot,
I tried to put filterNode logic into new codes, but failed
@MatusKysel please help recover the `filterNode` feature